### PR TITLE
Add support for more complicated documentation examples

### LINKED ISSUE #7143: Add a configuration file in documentation example for message
### Current problem

Right now we do not show the configuration used for example. We could and it would:

- show readers what need to be done to activate the check if it's an extension, or a standard but optional check (soon ™️)
- permits to create simpler examples by lowering the values to something sane (i.e. max line length = 15 instead of having 120+ character in the example, max-return-values = 2 instead of 11...)
- demonstrate some options that can affect the message (!)

### Desired solution

Show the configuration used in bad/good and use it as an example of configuration related to a message. There's some issue with this:

- How to identify related options?

It's not easy to determine that without reading the code, we struggle as much as user here.

- How to show the difference between the default value and the value being used in a test?

If the default value is 120 we need a way to show what it is if we change it to 15.

- How to handle duplication of the load-plugins warning

We already have a message for extensions right now.

### Additional context

Initially discussed in https://github.com/PyCQA/pylint/pull/7142#pullrequestreview-1031209077

### PR DESCRIPTION
**Pull request by DanielNoord**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :scroll: Docs          |

## Description

Refs https://github.com/PyCQA/pylint/issues/7897
Closes #7143

### CODE REVIEW

File: doc/data/messages/c/cyclic-import/bad/bad.py [hunk 1]
```diff
@@ -0,0 +1,9 @@
+from .bad2 import count_to_two  # [cyclic-import]
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/c/cyclic-import/bad/bad.py:1]: ~~Can we take into account the fact that if we want to do multiple examples properly displayed we're going to want to have bad.py, bad2.py, bad3.py (...) too ? Maybe the display can change if there's no good2.py, (...)  matching the bad2.py (...) ?~~ Edit: Did not notice it's in a directory, I think it's compatible.

File: pylint/checkers/imports.py [hunk 2]
```diff
@@ -555,13 +555,18 @@ def visit_importfrom(self, node: nodes.ImportFrom) -> None:
...
+        if imported_module_name is None:
             return
+
         for name, _ in node.names:
             if name != "*":
-                self._add_imported_module(node, f"{imported_module.name}.{name}")
+                self._add_imported_module(node, f"{imported_module_name}.{name}")
             else:
-                self._add_imported_module(node, imported_module.name)
+                self._add_imported_module(node, imported_module_name)
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/imports.py:569]: Seems like something that could be it's own PR.

**Review by Pierre-Sassoulas — COMMENTED** [review]
Looks pretty good, thank you for tackling this !

File: doc/test_messages_documentation.py [hunk 3]
```diff
@@ -52,10 +52,20 @@ def get_functional_test_files_from_directory(input_dir: Path) -> list[tuple[str,
                 suite.append(
                     (message_dir.stem, message_dir / "good.py"),
                 )
+
+            # Bad files can either consist of a single file or a directory.
             if (message_dir / "bad.py").exists():
                 suite.append(
                     (message_dir.stem, message_dir / "bad.py"),
                 )
+            elif (message_dir / "bad").is_dir():
```
[the remark this answers is not in the document]
Comment (DanielNoord) [reply doc/test_messages_documentation.py:61]: I think in that case it makes more sense to just use one file

File: pylint/checkers/imports.py (hunk 2 above)
Comment (DanielNoord) [reply pylint/checkers/imports.py:569]: Yeah and it is breaking tests... Without it this test is very flaky so I'm not sure what to do

**Comment by DanielNoord** [comment]
@Pierre-Sassoulas I kept in the `cyclic-import` example to show this works. I'll remove it afterwards as the behaviour is flaky and it seems to be because of `astroid`.

I think this PR also closes https://github.com/PyCQA/pylint/issues/7143 right?

File: doc/test_messages_documentation.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (ollie-iterators) [reply doc/test_messages_documentation.py:61]: You could also do bad.py, neutral.py, lawful.py and good.py

**Comment by Pierre-Sassoulas** [comment]
> I think this PR also closes https://github.com/PyCQA/pylint/issues/7143 right?

Yes, amazing !

File: doc/exts/pylint_messages.py [hunk 4]
```diff
@@ -62,28 +59,67 @@ def _register_all_checkers_and_extensions(linter: PyLinter) -> None:
...
-        title="Problematic code", text=_get_python_code_as_rst(bad_py_path)
-    )
+    # Handle multi-file bad code templates
+    if bad_py_path.exists():
+        bad_code = _get_titled_rst(
+            title="Problematic code", text=_get_python_code_as_rst(bad_py_path)
+        )
+    elif (data_path / "bad").exists():
+        bad_one = f"""\
+File one:
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_messages.py:85]: Instead of ``File one`` / ``File two``, we should have the actual name of the file (``bad.py::``. This would also simplify the handling code that can become generic and handle as many file as we want. We can also name it however we want (it's the directory prefix that indicate badness / goodness).

Comment (DanielNoord) [reply doc/exts/pylint_messages.py:85]: I like this suggestion! Will (hopefully) work on this tomorrow!

File: doc/exts/pylint_messages.py [hunk 5]
```diff
@@ -62,28 +59,62 @@ def _register_all_checkers_and_extensions(linter: PyLinter) -> None:
...
+``{file.name}``:
+
+.. literalinclude:: /{file.relative_to(Path.cwd())}
+    :language: python
+
+"""
+                )
+        bad_code = _get_titled_rst(title="Problematic code", text="".join(bad_files))
+    else:
+        bad_code = ""
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_messages.py:98]: We should create a generic function for this as we also need to handle ``good/`` directories with multiple file imo. I can do it, it's a lot easier with pycharm than vscode.

File: doc/exts/pylint_messages.py [hunk 6]
```diff
@@ -62,28 +59,62 @@ def _register_all_checkers_and_extensions(linter: PyLinter) -> None:
...
-        title="Problematic code", text=_get_python_code_as_rst(bad_py_path)
-    )
+    # Handle multi-file bad code templates
+    if bad_py_path.exists():
+        bad_code = _get_titled_rst(
+            title="Problematic code", text=_get_python_code_as_rst(bad_py_path)
+        )
+    elif (data_path / "bad").exists():
+        bad_files: list[str] = []
+        for file in (data_path / "bad").iterdir():
```
Comment (Pierre-Sassoulas) [inline doc/exts/pylint_messages.py:85]: We need to sort here, so bad.py is before bad2.py
```suggestion
        for file in sorted((data_path / "bad").iterdir()):
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
Let me know if you agree with the good/ multiple fle and I'll do it :)

File: doc/exts/pylint_messages.py (hunk 5 above)
Comment (DanielNoord) [reply doc/exts/pylint_messages.py:98]: Yeah I didn't want to add code that we don't actively use right now, but go ahead and add it! Whatever makes it easier to complete this process the quickest!

File: doc/exts/pylint_messages.py (hunk 6 above)
Comment (Pierre-Sassoulas) [reply doc/exts/pylint_messages.py]: Turn out it's not that easy to sort Path 😄 took me embarrassingly long to understand

**Comment by DanielNoord** [comment]
I have commits almost ready that clean up this PR. Hopefully can finish this tomorrow.

**Comment by DanielNoord** [comment]
Not clearing that reporter after every file shows some underlying issues that we never seen. Nice to fix these as well I think :smile:

File: pylint/testutils/reporter_for_tests.py [hunk 7]
```diff
@@ -72,9 +72,6 @@ def _display(self, layout: Section) -> None:
 
 
 class FunctionalTestReporter(BaseReporter):
-    def on_set_current_module(self, module: str, filepath: str | None) -> None:
-        self.messages = []
-
```
Comment (Pierre-Sassoulas) [inline pylint/testutils/reporter_for_tests.py:77]: Nice !

**Review by Pierre-Sassoulas — COMMENTED** [review]
LGTM but it's impossible to check in read the doc now that the cyclic-import is removed.

**Comment by DanielNoord** [comment]
> LGTM but it's impossible to check in read the doc now that the cyclic-import is removed.

Yeah, like I said; the `cyclic-import` message is 100% flakey 😅 

If you know of another message where you want multiple files and that isn't flakey I can add it as well. Otherwise I think we should just merge this. I did test this locally with multiple files and checked that the tests still worked (which is how I found the issues with the reporter).

**Comment by Pierre-Sassoulas** [comment]
Maybe too many lines ? It's a little insulting to readers though.  Otherwise, duplicated code ? I can add this and test locally myself :)

**Comment by DanielNoord** [comment]
> Maybe too many lines ? It's a little insulting to readers though. Otherwise, duplicated code ? I can add this and test locally myself :)

Duplicated code would be a good one! Perhaps make a PR that has this branch as target branch?

**Review by Pierre-Sassoulas — APPROVED** [review]
I'll let you check the example, and then let's merge. The result is pretty good, excited to make this live ! 😄

**Comment by DanielNoord** [comment]
Both https://pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/warning/while-used.html and https://pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/refactor/duplicate-code.html are definite improvements!

Indeed happy we have this now!

**Comment by stdedos** [comment]
> Both [pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/warning/while-used.html](https://pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/warning/while-used.html) and [pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/refactor/duplicate-code.html](https://pylint--8287.org.readthedocs.build/en/8287/user_guide/messages/refactor/duplicate-code.html) are definite improvements!
> 
> Indeed happy we have this now!

btw - is the `Similar lines in %s files %s` bug or feature?
Since it's "technically code", should the messages have the `backtick` formatting?

**Comment by DanielNoord** [comment]
I think it might make sense indeed!
